Tensordot support

anaruse opened the pull request:
This PR is to support tensordot as Chainer's function. You can use this function like numpy tensordot as below. Off course, it is backprop-able.

    import chainer
    import numpy
    a = chainer.Variable(numpy.random.rand(4,3,2))
    b = chainer.Variable(numpy.random.rand(3,2,5))
    c = chainer.functions.tensordot(a, b, axes=((1,2), (0,1)))

hvy commented:
Sorry for my late response, and thank you for this PR! It looks like tests are already in-place but were there uncovered cases that you were aware of? Also, it'd be great if you could add documentations for the implemented FunctionNode as discussed earlier and I think we can proceed.

anaruse commented:
@hvy: Thank you for your response.

I've just added docstrings to `tensordot` (I borrowed most sentences from cupy.tensordot).
Regarding tests, there are tests already and I'd like you to check if the coverage is enough or not.
Anyway, please review it!

hvy commented:
@anaruse Thank you for the clarification. I'll check accordingly!

hvy commented:
jenkins, test this please.

hvy commented:
Could you add reference to the docs (https://github.com/anaruse/chainer/blob/tensordot_support/docs/source/reference/functions.rst) so that this function is listed among the other ones?

anaruse commented:
Thank you for your comment. I've added the reference to the docs.

hvy commented:
jenkins, test this please.

hvy commented on chainer/functions/math/tensordot.py:
```diff
@@ -0,0 +1,171 @@
+            self.c_axes = c_axes
+
+        c = _tensordot(a, b, self.a_axes, self.b_axes, self.c_axes)
+        if self.dtype is not None and c.dtype != self.dtype:
```
It is possible to skip this conditioning and let `utils.force_array(c, self.dtype)` take care of it?

anaruse replied:
Thank you for your suggestion. I didn't know that `utils.force_array` has `dtype` option.

hvy commented on chainer/functions/math/tensordot.py:
```diff
@@ -0,0 +1,171 @@
+        elif isinstance(axes, int):
+            pass
+        else:
+            raise TypeError('axes must be int or tuple of int')
```
What do you think of making the error message slightly more verbose?

hvy commented on tests/chainer_tests/functions_tests/math_tests/test_tensordot.py:
```diff
@@ -0,0 +1,141 @@
+        self.forward_answer = self._get_forward_answer(self.a, self.b,
+                                                       self.axes)
+
+    def _get_forward_answer(self, a, b, axes):
```
I see that this method is aligned to other tests but I don't think it's needed for tensor dot (since the variable name is descriptive and there are no branches).

hvy commented:
As for tests, how about covering the following cases.

- Shapes with ndim 0, 1, and 2 (with axes that are valid for corresponding shapes. Maybe ndim 0 doesn't have to be supported?)
- 'axes:' 0
- default axes (maybe this isn't needed)
- Invalid axes, e.g. negative value or larger tuples

hvy commented on chainer/functions/math/tensordot.py:
```diff
@@ -0,0 +1,171 @@
+            a_type.dtype.kind == 'f',
+            b_type.dtype.kind == 'f',
+            a_type.ndim >= 1,
+            b_type.ndim >= 1,
```
Is ndim of 0 too trivial? Since numpy/cupy support it, I think I'd be nice if Chainer could too.

anaruse replied:
Agree. will check if it works well when ndim is zero.

anaruse replied:
I'm wondering how to handle an input with ndim=0, mainly because `chainer.Variable` does not accepts a scalar as its input. If you make it ndarray with shape of (1), then it becomes ndim=1.
Well, do we really need to consider the case of ndim=0?

hvy replied:
Ah, I think you can create a non-scalar ndarray with ndim=0 using e.g. 

```
x_data = numpy.random.uniform(-1, 1, ()))
x = chainer.Variable(x)
```

I am not sure how important this feature is but I noticed that e.g. matmul does support this case. Does it require a lot of changes to the code? https://github.com/chainer/chainer/blob/master/chainer/functions/math/matmul.py#L83

anaruse replied:
Thank you!  I didn't know how to make a ndarray with ndim=0. I don't think it requires a lot of changes to support it. will fix soon.

hvy commented:
It LGTM except some comments. Could you please take a look?

anaruse commented:
Thank you for your detailed feedback, @hvy!

I've fixed the branch based on your suggestion and added several tests for wider coverage (except ndim=0 case).
Please review again!

hvy commented:
Thanks for the quick response. Could you please take another look at my comment?

anaruse commented:
Thank you for your feedback. I've just fixed the branch. Now it support a ndarray with ndim=0.

anaruse commented:
I've just updated the branch so that it skips processing for transposing output tensor when it is obviously not necessary.

hvy commented on chainer/functions/math/tensordot.py:
```diff
@@ -0,0 +1,175 @@
+            raise ValueError('shape mismatch')
+
+    if a.ndim == 0 or b.ndim == 0:
+        y = a * b
```
Is this for performance reasons or does `xp.tensordot` in this case not work?

anaruse replied:
I thought `xp.tensordot` did not work when ndim=0, but seems it was because `list` was used for `axes` option at that time. It works OK if tuple is used. will use `xp.tensordot` even when ndim=0.

hvy commented on tests/chainer_tests/functions_tests/math_tests/test_tensordot.py:
```diff
@@ -0,0 +1,183 @@
+
+        {'a_shape': (2), 'b_shape': (3), 'axes': 0, 'gc_shape': (2, 3)},  # NOQA
+        {'a_shape': (2, 3), 'b_shape': (), 'axes': 0, 'gc_shape': (2, 3)},  # NOQA
+        {'a_shape': (), 'b_shape': (2, 3), 'axes': 0, 'gc_shape': (2, 3)},  # NOQA
```
How about testing `{'a_shape': (), 'b_shape': (),...},` too.

anaruse replied:
OK, will add that test too.

hvy commented on chainer/functions/math/tensordot.py:
```diff
@@ -0,0 +1,175 @@
+
+
+def _tensordot(a, b, a_axes, b_axes, c_axes=None):
+
```
Quite a nitpick but I think we can remove this empty line!

anaruse replied:
All right, will remote that empty line :)

hvy commented:
Sorry for my delayed response. I added some small comments. Do you mind taking another look?

anaruse commented:
Thank you for your feedback. I've just updated the branch based on your feedback. Please review again!

hvy commented:
jenkins, test this please.

hvy commented:
LGTM! Thank you again for this PR and all the quick fixes!
